x-pack/plugins/elastic_assistant/server/ai_assistant_data_clients/knowledge_base/helpers.ts

```diff
@@ -14,3 +16,87 @@ export const isModelAlreadyExistsError = (error: Error) => {
       error.body.error.type === 'status_exception')
   );
 };
+
+/**
+ * Returns an Elasticsearch query DSL that performs a vector search against the Knowledge Base for the given query/user/filter.
+ *
+ * @param filter - Optional filter to apply to the search
+ * @param kbResource - Specific resource tag to filter for, e.g. 'esql' or 'user'
+ * @param modelId - ID of the model to search with, e.g. `.elser_model_2`
+ * @param query - The search query provided by the user
+ * @param required - Whether to only include required entries
+ * @param user - The authenticated user
+ * @returns
+ */
+export const getKBVectorSearchQuery = ({
+  filter,
+  kbResource,
+  modelId,
+  query,
+  required,
+  user,
+}: {
+  filter?: QueryDslQueryContainer | undefined;
+  kbResource?: string | undefined;
+  modelId: string;
+  query: string;
+  required?: boolean | undefined;
+  user: AuthenticatedUser;
+}): QueryDslQueryContainer => {
+  const resourceFilter = kbResource
+    ? [
+        {
+          term: {
+            'metadata.kbResource': kbResource,
+          },
+        },
+      ]
+    : [];
+  const requiredFilter = required
+    ? [
+        {
+          term: {
+            'metadata.required': required,
+          },
+        },
+      ]
+    : [];
+
+  const userFilter = [
+    {
+      nested: {
```

Comment:
I think we should extend the filter to support empty("shared") KBs, which are available for all users in the space.

Reply:
I'm working through adding the different permutations of kb mock data today for https://github.com/elastic/kibana/pull/184974 and will use that to start crafting tests to cover these scenarios. 

I want to take a moment and rethink how we're categorizing/namespacing content using `kbResource` though to see if there will be any issues here. For now I've kept things matching the original KB implementation, and just introduced the `user` kbResource for all user created entries. We will also need to capture `entryType` somewhere to differentiate between raw text content and index-backed entries. Would be nice to have a catch-all `tags` for organization and labeling too. 

I will start thinking through more of this early next week as I get back to plumbing through the remainder of the kbDataClient methods and REST API's, but if you have any thoughts here I would love to hear them 🙂